segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/ChunkedSegmentStorage.java

```diff
@@ -1045,7 +1059,7 @@ private void concatUsingAppend(ConcatArgument[] concatArgs) throws ChunkStorageE
                 }
 
                 // Finally commit.
-                txn.commit(chunksToDelete.size() == 0); // if layout did not change then commit with lazyWrite.
+                txn.commit();
```

Comment:
I don't understand this change, can you elaborate, please?

Reply:
This is a fix for a bug that I found during the testing.

Quick background - 
When we commit using lazyCommit, we do not write changes to the table store immediately, but instead keep updated records in the memory and keep reusing/updating these in memory copies. This is done to avoid updating records when we are appending to the same chunk again and again. Such write operation causes chunk length and  segment length in metadata to change but otherwise nothing else changes. 
Now, in case of fail over this data stored in memory is lost.  But this very specific information can be reconstructed by looking at the actual chunk length and adjusting the metadata while doing `openRead` or `openWrite`.
For any other changes (addition/deletion of chunks, metadata updates like sealed etc) we must commit the updated record to the underlying store immediately because this change is not reconstructible.

What was the bug ?
I was incorrectly handling the "corner case"  where segment is truncated but first chunk is not removed. In this case only the start offset recorded in metadata changes but nothing else changes.
I was incorrectly committing with lazyCommit for truncate operation thinking if the layout (i.e chunks in the segment) did not change then we don't have to commit eagerly.
This is incorrect because in case of fail over the start offset changes but will be lost and this change is not recoverable.

What is the fix ?
Always commit eagerly as it is required in all cases.